datafusion/src/physical_plan/expressions/nth_value.rs

```diff
@@ -135,8 +135,12 @@ impl BuiltInWindowFunctionExpr for NthValue {
             NthValueKind::Last => (num_rows as usize) - 1,
             NthValueKind::Nth(n) => (n as usize) - 1,
         };
-        let value = ScalarValue::try_from_array(value, index)?;
-        Ok(value.to_array_of_size(num_rows))
+        Ok(if index >= num_rows {
+            new_null_array(value.data_type(), num_rows)
+        } else {
+            let value = ScalarValue::try_from_array(value, index)?;
+            value.to_array_of_size(num_rows)
```

Comment:
The same here applies for normal aggregations as probably happens here: if we have a `partition by` that creates a lot of groups, we will create many individual arrow arrays (which is slow / memory consuming).

Probably what would be better in the long run is store the offsets to the values in a contiguous array, and the values as well and extend / update them instead.

Reply:
Not needed for this PR btw, but just noting there are similar needs/performance issues in both aggregation and window functions.

Reply:
I agree, not in this pull request but I believe this can warrant a dedicated compute kernel in arrow for batched array slice transformation and then concatenation

Reply:
@Dandandan down the road I've started to work on this issue:

https://github.com/apache/arrow-datafusion/pull/579/files#diff-8b6b5ea3976c91229244e4e7a31a7026422b1374d1683e44b41af67a6bd43187R246-R254

```diff
-        let results = partition_points
-            .iter()
-            .map(|partition_range| {
-                let sort_partition_points =
-                    find_ranges_in_range(partition_range, &sort_partition_points);
-                let mut window_accumulators = self.create_accumulator()?;
-                sort_partition_points
-                    .iter()
-                    .map(|range| window_accumulators.scan_peers(&values, range))
-                    .collect::<Result<Vec<_>>>()
-            })
-            .collect::<Result<Vec<Vec<ArrayRef>>>>()?
-            .into_iter()
-            .flatten()
-            .collect::<Vec<ArrayRef>>();
-        let results = results.iter().map(|i| i.as_ref()).collect::<Vec<_>>();
-        concat(&results).map_err(DataFusionError::ArrowError)
+        let mut result = Vec::with_capacity(num_rows);
+        for partition_range in partition_points {
+            let sort_partition_points =
+                find_ranges_in_range(&partition_range, &sort_partition_points);
+            let mut window_accumulators = self.create_accumulator()?;
+            for range in sort_partition_points {
+                result.extend(window_accumulators.scan_peers(&values, range)?);
+            }
+        }
+        ScalarValue::iter_to_array(result.into_iter())
```

Reply:
Yes - that should probably already be quite an improvement :+1: